feat: uniffi bindings

jackstar12 opened the pull request:
TODO:
- [x] README for python lib and bindings crate
- [x] tests - ideally figure out how to share code with examples

michael1011 commented on bindings/.python-version:
```diff
@@ -0,0 +1 @@
+3.13
```
That's *really* new. What's the min Python version for the bindings?

michael1011 commented on bindings/pyproject.toml:
```diff
@@ -0,0 +1,16 @@
+version = "0.1.0"
+description = "Add your description here"
+readme = "README.md"
+requires-python = ">=3.10"
```
Nevermind

michael1011 commented on bindings/test.py:
```diff
@@ -0,0 +1,88 @@
+from datetime import datetime
+
+
+async def main():
```
This file looks less like a test more like example code

jackstar12 replied:
It is

michael1011 reviewed (commented):
Concept ACK

kilrau commented:
Let's fix linting and todo's and get this one over the finish line :muscle:

michael1011 commented on .gitignore:
```diff
@@ -4,3 +4,7 @@ boltz-*json
+
+# Bindings
+__pycache__/
+
```
Rm empty line

michael1011 commented on bindings/Cargo.toml:
```diff
@@ -0,0 +1,28 @@
+edition = "2021"
+
+[dependencies]
+elements = "0.25.0"
+thiserror = "1.0.50"
+uniffi = { version = "0.29.2", features = ["cli", "tokio"] }
+serde_json = "1"
+boltz-client = { path = "..", features = ["ws", "esplora", "electrum"] }
+bitcoin = "0.32.5"
+async-trait = "0.1.88"
+tokio = "1.44.1"
```
Let's inherit shared dependencies from the workspace. Like this: https://github.com/BoltzExchange/boltz-backend/blob/master/boltzr-cli/Cargo.toml#L9

michael1011 commented on bindings/python/.gitignore:
```diff
@@ -0,0 +1,3 @@
+__pycache__
+*.so
+boltz_client.py
```
Missing EOF empty line

michael1011 commented on bindings/README.md:
```diff
@@ -0,0 +1,11 @@
+To build the python bindings, run `cargo build --release` in the root directory.
+
+To run the tests, run `cargo test --release` in the root directory.
+
```
Rm duplicate empty line at the end of the file

michael1011 commented on bindings/README.md:
```diff
@@ -0,0 +1,11 @@
+# Bindings
+
+Uniffi bindings for the Boltz Rust library.
+Only python bindings are implemented in this repository, but uniffi potentially supports many other languages.
```
```suggestion
Only Python bindings are implemented in this repository, but uniffi supports many other languages.
```

michael1011 commented on bindings/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## Python
+
+To build the python bindings, run `cargo build --release` in the root directory.
```
```suggestion
To build the Python bindings, run `cargo build --release` in the root directory.
```

michael1011 commented on bindings/README.md:
```diff
@@ -0,0 +1,11 @@
+
+To build the python bindings, run `cargo build --release` in the root directory.
+
+To run the tests, run `cargo test --release` in the root directory.
```
Do you usually run the tests in release mode? I never did that

michael1011 commented on bindings/python/examples/chain.py:
```diff
@@ -0,0 +1,169 @@
+            claim_tx = await claim_script.construct_claim(preimage, claim_params)
+            print("Claim transaction constructed, broadcasting...")
+            await chain_client.broadcast_tx(claim_tx)
+            print("Claim transaction broadcasted!")
```
```suggestion
            print("Claim transaction broadcast!")
```

Not sure if "broadcasted" is s word

michael1011 commented on bindings/python/examples/chain.py:
```diff
@@ -0,0 +1,169 @@
+            refund_tx = await lockup_script.construct_refund(refund_params)
+            print("Refund transaction constructed, broadcasting...")
+            await refund_params.chain_client.broadcast_tx(refund_tx)
+            print("Refund transaction broadcasted!")
```
```suggestion
            print("Refund transaction broadcast!")
```

michael1011 commented on bindings/python/examples/submarine.py:
```diff
@@ -0,0 +1,112 @@
+    key_pair = boltz_client.KeyPair()
+
+    # Create a submarine swap request
+    # Note: Replace this with your actual Lightning invoice
```
```suggestion
```
They enter their lightning invoice

michael1011 commented on bindings/python/pyproject.toml:
```diff
@@ -0,0 +1,14 @@
+[project]
+name = "boltz_client"
+version = "0.1.0"
+description = "Add your description here"
```
```suggestion
version = "1.0.0"
description = "Boltz Swap library"
```

michael1011 commented on bindings/src/boltz.rs:
```diff
@@ -0,0 +1,516 @@
+    pub lockup_details: ChainSwapDetails,
+}
+
+/*
+fn check_limits_within(maximal: u64, minimal: u64, output_amount: u64) -> Result<(), Error> {
+    if output_amount < minimal {
+        return Err(Error::Protocol(format!(
+            "Output amount is below minimum {minimal}"
+        )));
+    }
+    if output_amount > maximal {
+        return Err(Error::Protocol(format!(
+            "Output amount is above maximum {maximal}"
+        )));
+    }
+    Ok(())
+}
+*/
```
?

michael1011 commented on bindings/tests/bindings/chain.py:
```diff
@@ -0,0 +1,88 @@
+from common import *
+
+async def swap(from_chain: boltz_client.Chain, to_chain: boltz_client.Chain):
+
```
```suggestion
```

michael1011 commented on bindings/tests/bindings/common.py:
```diff
@@ -0,0 +1,58 @@
+            raise TimeoutError(f"Timeout waiting for status '{status}'")
+        print("Waiting for status", update.status)
+        if update.status == status:
+            return update
```
Missing empty line at EOF

michael1011 commented on bindings/tests/bindings/submarine.py:
```diff
@@ -0,0 +1,49 @@
+    await swap(lbtc_chain)
+
+if __name__ == "__main__":
+    asyncio.run(main())
```
Missing empty line at EOF of file

michael1011 commented on src/swaps/boltz.rs:
```diff
@@ -975,9 +985,6 @@ pub struct SwapStatus {
     #[serde(rename = "failureDetails", skip_serializing_if = "Option::is_none")]
     pub failure_details: Option<FailureReasonIncorrectAmounts>,
-
-    #[serde(rename = "channel", skip_serializing_if = "Option::is_none")]
-    pub channel_info: Option<ChannelInfo>,
```
Was that the only usage of `ChannelInfo`? If so, remove the struct too

jackstar12 replied:
isnt that legacy channel creation stuff? heres its content:
```
#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
pub struct ChannelInfo {
    #[serde(rename = "fundingTransactionId")]
    pub funding_transaction_id: String,
    #[serde(rename = "fundingTransactionVout")]
    pub funding_transaction_vout: u64,
}
```

michael1011 reviewed (commented):
`bindings/python/README.md` seems to be an empty file. Can we delete that one?

Do you want to use ruff as formatter for the Python code?

jackstar12 committed:
feat: uniffi bindings

jackstar12 committed:
refactor: use workspace dependencies

jackstar12 committed:
ci: setup vu in lint workflow

jackstar12 committed:
chore: formatting

jackstar12 committed:
refactor: run bindings tests in serial

jackstar12 committed:
chore: add delay to submarine swap tests

lightning node can get out of sync otherwise because of all the blocks
being mined

michael1011 commented on bindings/tests/bindings/common.py:
```diff
@@ -0,0 +1,72 @@
+
+
+async def delay():
+    await asyncio.sleep(5)
```
Not a fan of just sleeping. We should actually check if the nodes are in sync
But I guess this should be fine for now

michael1011 commented on bindings/Makefile:
```diff
@@ -0,0 +1,23 @@
+	cd python && uv build
+
+format-python:
+    # also format python tests
```
```suggestion
	# Also formats the Python tests
```

michael1011 commented on bindings/README.md:
```diff
@@ -0,0 +1,65 @@
+
+Uniffi bindings for the Boltz Rust library.
+
+[UniFFI](https://mozilla.github.io/uniffi-rs/) enables automatic generation of bindings for multiple programming languages from a single Rust codebase. Currently, only Python bindings are implemented in this repository, but UniFFI potentially supports many other languages including Kotlin, Swift, Ruby, and more.
```
```suggestion
[UniFFI](https://mozilla.github.io/uniffi-rs/) enables automatic generation of bindings for multiple programming languages from a single Rust codebase. Currently, only Python bindings are implemented in this repository, but UniFFI supports many other languages including Kotlin, Swift, Ruby, and more.
```

michael1011 committed:
chore: minor wording cleanup

i5hi commented on bindings/src/swap.rs:
```diff
@@ -0,0 +1,205 @@
+pub struct SwapScript(swaps_bitcoin::SwapScript);
+
+#[derive(uniffi::Record)]
+pub struct SwapTransactionParams {
```
Why is Arc used on some fields and not others?

michael1011 replied:
`uniffi::Object`s need to be wrapped in `Arc` to be able to use export bindings to them

i5hi commented on bindings/src/network.rs:
```diff
@@ -0,0 +1,133 @@
+                        electrum.validate_domain,
+                        electrum.timeout,
+                    )
+                    .unwrap(),
```
Rather than unwrap would be better to either return a Result or an Option<Self>

i5hi commented:
L142 in bindings/boltz.rs

```
preimage_hash: swap_request.preimage_hash.parse::<sha256::Hash>().unwrap(),
```

We should be able to just use `?` here since the function returns a Result

i5hi commented:
network.rs

L100 & L119 unwrap on Electrum Client creation

i5hi commented:
I would suggest to do a codebase wide search of `unwrap()` and make sure they are handled gracefully.

michael1011 committed:
chore: remove unwraps

michael1011 commented:
@i5hi I removed the unwraps from the code touched in this PR. Getting rid of them in the entire codebase should be done separately imho
